File: Source/BackupFiles/NARAE/SMClient/MultiRBTree.cpp
#include "MultiRBTree.h"
#include  <string.h>


int CompareStr(char* Item1, char* Item2)
{

	//return (strncmp(Item1, Item2, len));
	return (strcmp(Item1, Item2));
}

//TMultiRBTree::TMultiRBTree(int nKeyLen)
//{
//	CompareFunc = CompareStr;
//	Root = NULL;
//	Leftmost = NULL;
//	Rightmost = NULL;
//	FKeyLen = nKeyLen;
//}


TMultiRBTree::TMultiRBTree()
{
	CompareFunc = CompareStr;
	Root = NULL;
	Leftmost = NULL;
	Rightmost = NULL;
	FKeyLen = 0;
}



TMultiRBTree::~TMultiRBTree()
{
	Clear();
}

void TMultiRBTree::SetKeyLen(int len)
{
	FKeyLen = len;
}


void TMultiRBTree::RBInc(TMultiRBNode* x)
{
	TMultiRBNode* y;

	if (x->Right != NULL)
	{
		x = x->Right;
		while (x->Left != NULL)
		{
			x = x->Left;
		}
	}
	else
	{
		y = x->Parent;
		while (x == y->Right)
		{
			x = y;
			y = y->Parent;
		}
		if (x->Right != y)
			x = y;
	}
}

//{ Pre: x != first }
void TMultiRBTree::RBDec(TMultiRBNode* x)
{
	TMultiRBNode* y;

	if (x->Left != NULL)
	{
		y = x->Left;
		while (y->Right != NULL)
		{
			y = y->Right;
		}
		x = y;
	}
	else
	{
		y = x->Parent;
		while (x == y->Left)
		{
			x = y;
			y = y->Parent;
		}
		x = y;
	}
}

void fast_erase(TMultiRBNode* x)
{
	if (x->Left != NULL)
		fast_erase(x->Left);
	if (x->Right != NULL)
		fast_erase(x->Right);

	delete x;
}

void TMultiRBTree::Clear()
{
	if (Root != NULL)
		fast_erase(Root);
	Root = NULL;
	Leftmost =  NULL;
	Rightmost = NULL;
}

TMultiRBNode* TMultiRBTree::Find(char* pKey)
{
	char Key[MAXLEN_KEY];
	//FORMAT_KEY(pKey, FKeyLen, Key);

	int cmp;
	TMultiRBNode* Result = Root;
	while (Result != NULL)
	{
		//cmp = CompareFunc(Result->Key, Key, FKeyLen);
		cmp = CompareFunc(Result->Key, Key);
		if (cmp < 0) {
		Result = Result->Right;
		}
		else if (cmp > 0) {
			Result = Result->Left;
		}
		else
		{
			break;
		}
	}
	return Result;
}

void TMultiRBTree::RotateLeft(TMultiRBNode* x)
{
	TMultiRBNode* y;

	y = x->Right;
	x->Right = y->Left;

	if (y->Left != NULL)
		y->Left->Parent = x;

	y->Parent = x->Parent;
	if (x == Root) {
		Root = y;
	}
	else if (x == x->Parent->Left) {
		x->Parent->Left = y;
	}
	else {
		x->Parent->Right = y;
	}

	y->Left = x;
	x->Parent = y;
}

void TMultiRBTree::RotateRight(TMultiRBNode* x)
{
	TMultiRBNode* y;

	y = x->Left;
	x->Left = y->Right;
	
	if (y->Right != NULL)
		y->Right->Parent = x;
	
	y->Parent = x->Parent;
	if (x = Root) {
		Root = y;
	}
	else if (x = x->Parent->Right) {
		x->Parent->Right = y;
	}
	else {
		x->Parent->Left = y;
	}
	y->Right  = x;
	x->Parent  = y;
}

TMultiRBNode* TMultiRBTree::Minimum(TMultiRBNode* x)
{
	TMultiRBNode* Result = x;
	while (Result->Left != NULL)
		Result = Result->Left;
	return Result;
}

TMultiRBNode* TMultiRBTree::Maximum(TMultiRBNode* x)
{
	TMultiRBNode* Result = x;
	while (Result->Right != NULL)
		Result = Result->Right;

	return Result;
}

TMultiRBNode* TMultiRBTree::Add(char* pKey, void* Data)
{
	char Key[MAXLEN_KEY];
	//FORMAT_KEY(pKey, FKeyLen, Key);

	TMultiRBNode *x, *y, *z, *zpp;
	int cmp;
	TMultiRBNode* Result;

	z = new TMultiRBNode;
	
	//{ Initialize fields in new node z }
	strcpy(z->Key, Key);
	z->Left 	= NULL;
	z->Right 	= NULL;
	z->Color 	= clRed;
	z->Duplicates = new TRBTree();
	z->Duplicates->Add( (void*)Data, (void*)Data);
	Result = z;

	//{ Maintain Leftmost and Rightmost nodes }
	if ((Leftmost == NULL) || (CompareFunc(Key, Leftmost->Key) < 0))
		Leftmost = z;

	if ((Rightmost == NULL) || (CompareFunc(Rightmost->Key, Key) < 0))
		Rightmost = z;

	//{ Insert node z }
	y = NULL;
	x = Root;
	while (x != NULL)
	{
		y = x;
		cmp = CompareFunc(Key, x->Key);
		if (cmp < 0) {
			x = x->Left;
		}
		else if (cmp > 0) {
			x = x->Right;
		}
		else
		{
			//{ Value already exists in tree. }
			x->Duplicates->Add(Data, Data);
			Result = x;

			delete z->Duplicates;
			delete z;
			return Result;
		}
	} // while (x != NULL)

	z->Parent = y;
	if (y = NULL)
	{
		Root = z;
	}
	else if (CompareFunc(Key, y->Key) < 0)
	{
		y->Left = z;
	}
	else
	{
		y->Right = z;
	}

	//{ Rebalance tree }
	while ((z != Root) && (z->Parent->Color = clRed))
	{
		zpp = z->Parent->Parent;
		if (z->Parent == zpp->Left)
		{
			y = zpp->Right;
			if ((y != NULL) && (y->Color = clRed))
			{
				z->Parent->Color = clBlack;
				y->Color = clBlack;
				zpp->Color = clRed;
				z = zpp;
			}
			else
			{
				if (z == z->Parent->Right)
				{
					z = z->Parent;
					RotateLeft(z);
				}

				z->Parent->Color = clBlack;
				zpp->Color = clRed;
				RotateRight(zpp);
			}
		} // if (z->Parent == zpp->Left)	
		else
		{
			y = zpp->Left;
			if ((y != NULL) && (y->Color == clRed))
			{
				z->Parent->Color = clBlack;
				y->Color = clBlack;
				zpp->Color = clRed;
				z = zpp;
			}
			else
			{
				if (z == z->Parent->Left)
				{
					z = z->Parent;
					RotateRight(z);
				}
				z->Parent->Color = clBlack;
				zpp->Color = clRed;
				RotateLeft(zpp);
			}
		}
	}// end of while ((z != Root) && (z->Parent->Color = clRed))
	Root->Color  = clBlack;

}

void TMultiRBTree::Delete(TMultiRBNode* Z, void* Data)
{
	TMultiRBNode *w, *x, *y, *x_parent;
	TColor tmpcol;

	if (Z == NULL)
		return;

	if (Z->Duplicates->Count > 1)
	{
		Z->Duplicates->Delete(Z->Duplicates->Find(Data));
		return;
	}

	y = Z;
	x = NULL;
	x_parent = NULL;

	if (y->Left == NULL)
	{
		//{ z has at most one non-null child. y = z. }
		x = y->Right;     //{ x might be null. }
	}
	else
	{
		if (y->Right == NULL)
		{
			//{ z has exactly one non-null child. y = z. }
			x = y->Left;    //{ x is not null. }
		}
		else
		{
			//{ z has two non-null children.  Set y to }
			y = y->Right;   //{   z's successor.  x might be null. }
			while (y->Left != NULL) {
				y = y->Left;
			}
		}
		x = y->Right;
	}


	if (y != Z)
	{
		//{ "copy y's sattelite data into z" }
		//{ relink y in place of z.  y is z's successor }
		Z->Left->Parent = y;
		y->Left = Z->Left;

		if (y != Z->Right)
		{
			x_parent = y->Parent;
			if (x != NULL)
			{
				x->Parent = y->Parent;
			}
			y->Parent->Left = x; //{ y must be a child of Left }
			y->Right = Z->Right;
			Z->Right->Parent = y;
		}
		else
		{
			x_parent = y;
		}

		if (Root == Z) {
			Root = y;
		}
		else if (Z->Parent->Left == Z) {
			Z->Parent->Left = y;
		}
		else {
			Z->Parent->Right = y;
		}
		y->Parent = Z->Parent;
		tmpcol = y->Color;
		y->Color = Z->Color;
		Z->Color = tmpcol;
		y = Z;

		//{ y now points to node to be actually deleted }
	} // if (y != Z)
	else
	{                        //{ y = z }
		x_parent = y->Parent;
		if (x != NULL)
		{
			x->Parent = y->Parent;
		}

		if (Root == Z)
		{
			Root = x;
		}
		else
		{
			if (Z->Parent->Left == Z)
			{
				Z->Parent->Left = x;
			}
			else
			{
				Z->Parent->Right = x;
			}
		}

		if (Leftmost == Z)
		{
			if (Z->Right == NULL)
			{	      //{ z->Left must be null also }
				Leftmost = Z->Parent;
			}
			else
			{
				Leftmost = Minimum(x);
			}
		}
		if (Rightmost == Z)
		{
			if (Z->Left == NULL)
			{       //{ z->Right must be null also }
				Rightmost = Z->Parent;
			}
			else
			{                     //{ x == z->Left }
				Rightmost = Maximum(x);
			}
		}
	}// else of if (y != Z)

	//{ Rebalance tree }
	if (y->Color == clBlack)
	{
		while ((x != Root) && ((x == NULL) || (x->Color == clBlack)))
		{
			if (x == x_parent->Left)
			{
				w = x_parent->Right;
				if (w->Color = clRed)
				{
					w->Color = clBlack;
					x_parent->Color = clRed;
					RotateLeft(x_parent);
					w = x_parent->Right;
				}
				if (((w->Left == NULL) || (w->Left->Color == clBlack)) && ((w->Right = NULL) || (w->Right->Color == clBlack)))
				{
					w->Color = clRed;
					x = x_parent;
					x_parent = x_parent->Parent;
				}
				else
				{
					if ((w->Right == NULL) || (w->Right->Color == clBlack))
					{
						w->Left->Color = clBlack;
						w->Color = clRed;
						RotateRight(w);
						w = x_parent->Right;
					}
					w->Color = x_parent->Color;
					x_parent->Color = clBlack;
					if (w->Right != NULL)
					{
						w->Right->Color = clBlack;
					}
					RotateLeft(x_parent);
					x = Root; //{ break; }
				}
			} // if (x == x_parent->Left)
			else
			{
				//{ same as above, with Right <-> Left. }
				w = x_parent->Left;
				if (w->Color == clRed)
				{
					w->Color = clBlack;
					x_parent->Color = clRed;
					RotateRight(x_parent);
					w = x_parent->Left;
				}
				if (((w->Right == NULL) || (w->Right->Color == clBlack)) && ((w->Left == NULL) || (w->Left->Color = clBlack)))
				{
					w->Color = clRed;
					x = x_parent;
					x_parent = x_parent->Parent;
				}
				else
				{
					if ((w->Left == NULL) || (w->Left->Color == clBlack))
					{
						w->Right->Color = clBlack;
						w->Color = clRed;
						RotateLeft(w);
						w = x_parent->Left;
					}
					w->Color = x_parent->Color;
					x_parent->Color = clBlack;
					if (w->Left != NULL)
					{
						w->Left->Color = clBlack;
					}
					RotateRight(x_parent);
					x = Root; //{ break; }
				}
			}// else of if (x == x_parent->Left)

		} // while ((x != Root) && ((x == NULL) || (x->Color == clBlack)))
		if (x != NULL)
		{
			x->Color = clBlack;
		}
	} // if (y->Color == clBlack)
	delete y->Duplicates;
	delete y;
}

File: Source/BackupFiles/NARAE/SMClient/MultiRBTree.pas
unit MultiRBTree;

interface

uses
  Classes, RBTree;

type
  TStrSortCompare = function (const Item1: String; const Item2: String): Integer;
  TMultiColor = (clRed, clBlack);

  TMultiRBNodeP = ^TMultiRBNode;
  TMultiRBNode = record
    Key: String;
    Left, Right, Parent: TMultiRBNodeP;
    Color: TMultiColor;
    Duplicates: TRBTree;
//    Data: Pointer;
  end;

  TMultiRBTree = class
	private
  	Root: TMultiRBNodeP;
    Leftmost: TMultiRBNodeP;
    Rightmost: TMultiRBNodeP;
    CompareFunc: TStrSortCompare;
    procedure RotateLeft(var x: TMultiRBNodeP);
    procedure RotateRight(var x: TMultiRBNodeP);
    function Minimum(var x: TMultiRBNodeP): TMultiRBNodeP;
    function Maximum(var x: TMultiRBNodeP): TMultiRBNodeP;
	public
  	constructor Create;
    destructor Destroy; override;
    procedure Clear;
    procedure Delete(Z: TMultiRBNodeP; Data: Pointer);
    function Find(Key: String): TMultiRBNodeP;
    function Add(Key: String; Data: Pointer): TMultiRBNodeP;
  public
    property First: TMultiRBNodeP read leftmost;
    property Last: TMultiRBNodeP read rightmost;
	end;

  procedure RBInc(var x: TMultiRBNodeP);  
implementation

procedure RBInc(var x: TMultiRBNodeP);
var
  y: TMultiRBNodeP;
begin
  if (x^.Right <> nil) then
  begin
    x := x^.Right;
    while (x^.Left <> nil) do
    begin
      x := x^.Left;
    end;
  end
  else
  begin
    y := x^.Parent;
    while (x = y^.Right) do
    begin
      x := y;
      y := y^.Parent;
    end;
    if (x^.Right <> y) then
      x := y;
  end
end;

{ Pre: x <> first }
procedure RBDec(var x: TMultiRBNodeP);
var
  y: TMultiRBNodeP;
begin
  if (x^.left <> nil) then
  begin
    y := x^.Left;
    while (y^.Right <> nil) do
    begin
      y := y^.Right;
    end;
    x := y;
  end
  else
  begin
    y := x^.Parent;
    while (x = y^.Left) do
    begin
      x := y;
      y := y^.Parent;
    end;
    x := y;
  end
end;

function CompareStr(const Item1: String; const Item2: String): Integer;
begin
  if (Item1 < Item2) then begin
    Result := -1;
  end
  else if (Item1 = Item2) then
  begin
    Result := 0;
  end
  else
  begin
    Result := 1;
  end
end;

constructor TMultiRBTree.Create;
begin
  inherited Create;
  CompareFunc := CompareStr;
  Root := nil;
  Leftmost  := nil;
  Rightmost := nil;
end;

destructor TMultiRBTree.Destroy;
begin
  Clear;
  inherited Destroy;
end;

procedure fast_erase(x: TMultiRBNodeP);
begin
  if (x^.Left <> nil) then  fast_erase(x^.Left);
  if (x^.Right <> nil) then fast_erase(x^.Right);
  x^.Duplicates.Free;
  dispose(x);
end;

procedure TMultiRBTree.Clear;
begin
  if (Root <> nil) then
    fast_erase(Root);
  Root := nil;
  Leftmost  := nil;
  Rightmost := nil;
end;

function TMultiRBTree.Find(Key: String): TMultiRBNodeP;
var
  cmp: integer;
begin
  Result := Root;
  while (Result <> nil) do
  begin
    cmp := CompareFunc(Result^.Key, key);
    if cmp < 0 then
    begin
      Result := Result^.right;
    end
    else if cmp > 0 then
    begin
      Result := Result^.left;
    end
    else
    begin
      break;
    end;
  end;
end;

procedure TMultiRBTree.RotateLeft(var x: TMultiRBNodeP);
var
  y: TMultiRBNodeP;
begin
  y := x^.right;
  x^.right := y^.left;
  if (y^.left <> nil) then
  begin
    y^.left^.parent := x;
  end;
  y^.parent := x^.parent;
  if (x = Root) then
  begin
    Root := y;
  end
  else if (x = x^.parent^.left) then
  begin
    x^.parent^.left := y;
  end
  else
  begin
    x^.parent^.right := y;
  end;
  y^.left := x;
  x^.parent := y;
end;

procedure TMultiRBTree.RotateRight(var x: TMultiRBNodeP);
var
  y: TMultiRBNodeP;
begin
  y := x^.left;
  x^.left := y^.right;
  if (y^.right <> nil) then
  begin
    y^.right^.parent := x;
  end;
  y^.parent := x^.parent;
  if (x = Root) then
  begin
    Root := y;
  end
  else if (x = x^.parent^.right) then
  begin
    x^.parent^.right := y;
  end
  else
  begin
    x^.parent^.left := y;
  end;
  y^.right := x;
  x^.parent := y;
end;

function TMultiRBTree.Minimum(var x: TMultiRBNodeP): TMultiRBNodeP;
begin
  Result := x;
  while (Result^.left <> nil) do
    Result := Result^.left;
end;

function TMultiRBTree.Maximum(var x: TMultiRBNodeP): TMultiRBNodeP;
begin
  Result := x;
  while (Result^.right <> nil) do
    Result := Result^.right;
end;

function TMultiRBTree.Add(Key: String; Data: Pointer): TMultiRBNodeP;
var
  x, y, z, zpp: TMultiRBNodeP;
  cmp: Integer;
begin
  z := New(TMultiRBNodeP);
  { Initialize fields in new node z }
  z^.Key 		:= key;
  z^.Left 	:= nil;
  z^.Right 	:= nil;
  z^.Color 	:= clRed;
  z^.Duplicates := TRBTree.Create;
  z^.Duplicates.Add(data, data);
  Result := z;
  
  { Maintain leftmost and rightmost nodes }
  if ((leftmost = nil) or (compareFunc(key, leftmost^.Key) < 0)) then
  begin
    leftmost := z;
  end;

  if ((rightmost = nil) or (compareFunc(rightmost^.Key, key) < 0)) then
  begin
    rightmost := z;
  end;
  
  { Insert node z }
  y := nil;
  x := Root;
  while (x <> nil) do
  begin
    y := x;
    cmp := compareFunc(key, x^.Key);
    if (cmp < 0) then
    begin
      x := x^.left;
    end
    else if (cmp > 0) then
    begin
      x := x^.right;
    end
    else
    begin
      { Value already exists in tree. }
      x^.Duplicates.Add(data, data);
      Result := x;

      z^.Duplicates.Free;
      dispose(z);
      exit;
    end;
  end;

  z^.parent := y;
  if (y = nil) then
  begin
    Root := z;
  end
  else if (compareFunc(key, y^.Key) < 0) then
  begin
    y^.left := z;
  end
  else
  begin
    y^.right := z;
  end;

  { Rebalance tree }
  while ((z <> Root) and (z^.parent^.color = clRed)) do
	begin
  	zpp := z^.parent^.parent;
    if (z^.parent = zpp^.left) then
    begin
      y := zpp^.right;
      if ((y <> nil) and (y^.color = clRed)) then
      begin
        z^.parent^.color := clBlack;
        y^.color := clBlack;
        zpp^.color := clRed;
      	z := zpp;
      end
      else
      begin
        if (z = z^.parent^.right) then
        begin
          z := z^.parent;
          rotateLeft(z);
        end;
        z^.parent^.color := clBlack;
        zpp^.color := clRed;
        rotateRight(zpp);
      end;
    end
    else
    begin
      y := zpp^.left;
      if ((y <> nil) and (y^.color = clRed)) then
      begin
        z^.parent^.color := clBlack;
        y^.color := clBlack;
        zpp^.color := clRed;
        z := zpp;
      end
      else
      begin
        if (z = z^.parent^.left) then
        begin
          z := z^.parent;
          rotateRight(z);
        end;
        z^.parent^.color := clBlack;
        zpp^.color := clRed;
        rotateLeft(zpp);
      end;
    end;
  end;
  Root^.color := clBlack;
end;

procedure TMultiRBTree.Delete(Z: TMultiRBNodeP; Data: Pointer);
var
  w, x, y, x_parent: TMultiRBNodeP;
  tmpcol: TMultiColor;
begin
	if Z = NIL then Exit;

	if Z^.Duplicates.Count > 1 then
  begin
  	Z^.Duplicates.Delete(Z^.Duplicates.Find(Data));
		Exit;
  end;

  y := z;
  x := nil;
  x_parent := nil;

  if (y^.left = nil) then
  begin    { z has at most one non-null child. y = z. }
    x := y^.right;     { x might be null. }
  end
  else
  begin
    if (y^.right = nil) then
    begin { z has exactly one non-null child. y = z. }
      x := y^.left;    { x is not null. }
    end
    else
    begin
      { z has two non-null children.  Set y to }
      y := y^.right;   {   z's successor.  x might be null. }
      while (y^.left <> nil) do begin
        y := y^.left;
      end;
      x := y^.right;
    end;
  end;

  if (y <> z) then
  begin
    { "copy y's sattelite data into z" }
    { relink y in place of z.  y is z's successor }
    z^.left^.parent := y;
    y^.left := z^.left;
    if (y <> z^.right) then
    begin
      x_parent := y^.parent;
      if (x <> nil) then
      begin
        x^.parent := y^.parent;
      end;
      y^.parent^.left := x;   { y must be a child of left }
      y^.right := z^.right;
      z^.right^.parent := y;
    end
    else
    begin
      x_parent := y;
    end;
    if (Root = z) then
    begin
      Root := y;
    end
    else if (z^.parent^.left = z) then
    begin
      z^.parent^.left := y;
    end else begin
      z^.parent^.right := y;
    end;
    y^.parent := z^.parent;
    tmpcol := y^.color;
    y^.color := z^.color;
    z^.color := tmpcol;
    y := z;
    { y now points to node to be actually deleted }
  end
  else
  begin                        { y = z }
    x_parent := y^.parent;
    if (x <> nil) then
    begin
      x^.parent := y^.parent;
    end;
    if (Root = z) then
    begin
      Root := x;
    end
    else
    begin
      if (z^.parent^.left = z) then
      begin
        z^.parent^.left := x;
      end
      else
      begin
        z^.parent^.right := x;
      end;
    end;
	  if (leftmost = z) then
    begin
	    if (z^.right = nil) then
      begin      { z^.left must be null also }
	      leftmost := z^.parent;
	    end else
      begin
	      leftmost := minimum(x);
      end;
    end;
	  if (rightmost = z) then
    begin
	    if (z^.left = nil) then
      begin       { z^.right must be null also }
	      rightmost := z^.parent;
	    end
      else
      begin                     { x == z^.left }
	      rightmost := maximum(x);
      end;
    end;
  end;

  { Rebalance tree }
  if (y^.color = clBlack) then
  begin
    while ((x <> Root) and ((x = nil) or (x^.color = clBlack))) do
    begin
      if (x = x_parent^.left) then
      begin
          w := x_parent^.right;
          if (w^.color = clRed) then
          begin
            w^.color := clBlack;
            x_parent^.color := clRed;
            rotateLeft(x_parent);
            w := x_parent^.right;
          end;
          if (((w^.left = nil) or
               (w^.left^.color = clBlack)) and
              ((w^.right = nil) or
               (w^.right^.color = clBlack)))  then begin
            w^.color := clRed;
            x := x_parent;
            x_parent := x_parent^.parent;
          end
          else
          begin
            if ((w^.right = nil) or (w^.right^.color = clBlack)) then
            begin
              w^.left^.color := clBlack;
              w^.color := clRed;
              rotateRight(w);
              w := x_parent^.right;
            end;
            w^.color := x_parent^.color;
            x_parent^.color := clBlack;
            if (w^.right <> nil) then
            begin
              w^.right^.color := clBlack;
            end;
            rotateLeft(x_parent);
            x := Root; { break; }
         end
      end
      else
      begin
        { same as above, with right <^. left. }
        w := x_parent^.left;
        if (w^.color = clRed)  then begin
          w^.color := clBlack;
          x_parent^.color := clRed;
          rotateRight(x_parent);
          w := x_parent^.left;
        end;
        if (((w^.right = nil) or
             (w^.right^.color = clBlack)) and
            ((w^.left = nil) or
             (w^.left^.color = clBlack)))  then begin
          w^.color := clRed;
          x := x_parent;
          x_parent := x_parent^.parent;
        end else begin
          if ((w^.left = nil) or (w^.left^.color = clBlack)) then begin
            w^.right^.color := clBlack;
            w^.color := clRed;
            rotateLeft(w);
            w := x_parent^.left;
          end;
          w^.color := x_parent^.color;
          x_parent^.color := clBlack;
          if (w^.left <> nil) then begin
            w^.left^.color := clBlack;
          end;
          rotateRight(x_parent);
          x := Root; { break; }
        end;
      end;
    end;
    if (x <> nil) then begin
      x^.color := clBlack;
    end;
  end;
  y^.Duplicates.Free;
  dispose(y);
end;


end.

